### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive in while-loop with dynamic class property change

### ISSUE THREAD

**Issue by Ptitet** [typescript-eslint/typescript-eslint#9485]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=MYGwhgzhAEBCCuBzaBvAUG6XoFMAeAlgC4AyA9mQA7QC80AZmCBDgNwbbQC2YBAdgAoAlKkydswMnwhFoZIgAscAJwDCUgCbECU2tEUEIAOmVg%2BGsl1gUQw9uOwHj%2BYuSp7GzNmIcgcsyT4ifngcN2o6AEInIxdSCmoAMkS5RRV1c20pewdoAHcFAj9oAUDgvlDwkXRc8RieAE9VBTNEHABRQiIANSZQux9a6DKQsIS9aMLnLvDoZNSlNU0svhyHAF9BzcHG5taOrt6QfurBzhi42boY03NLazJbITXobc5biysbYVFa5X94Mo%2BNAALJgRQmMyfH4APmgAAYjABWF6bdZAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRgA2pETRoHaJAA0mcNhxD48DgHcAwpOT4hLADIcOxS9NnymqDPjiJ9EAL4Auvr%2BIP5AA&tsconfig=&tokens=false

### Repro Code

```TypeScript
let exitLoop = false;
const otherCondition = randomBool();

while (!exitLoop && otherCondition) {
    mayExit();
}

function mayExit() {
    exitLoop = randomBool();
}

function randomBool() {
    return Math.random() > 0.5;
}

// it also works with classes, it is how I discovered the bug:

class Bug {

    exitLoop = false;

    main() {
        const otherCondition = this.randomBool();
        this.exitLoop = false;
        while (!this.exitLoop && otherCondition) {
            this.mayExit();
        }
    }

    mayExit() {
        this.exitLoop = this.randomBool();
    }

    randomBool() {
        return Math.random() > 0.5;
    }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": ["error", {
      allowConstantLoopConditions: true
    }]
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected `!this.exitLoop` on line 8 not to be underlined in VSCode and not to be marked as an error according to the rule `@typescript-eslint/no-unnecessary-condition`.
It should not be underlined because the value of `exitLoop` can be changed by the methods called in the while-loop, so the condition is not constant and therefore needed.

### Actual Result

`!this.exitLoop` on line 8 is underlined in red and marked as an error according to the rule `@typescript-eslint/no-unnecessary-condition`.

### Additional Info

- TypeScript version: 5.5.3
- ESLint version: 9.6.0
- typescript-eslint version: 7.15.0

This bug does not happen when there is no `otherCondition`, as shown in the code samples I provided.

**Comment by kirkwaiblinger** [comment]
Hi @Ptitet!

This is expected. The rule is powered purely by the types, and TS treats exitLoop as `false`, since it doesn't track side effects of function calls.  See also, eg https://github.com/typescript-eslint/typescript-eslint/issues/9477.

Cheers
